### PR TITLE
feat(synonyms): exportSynonyms and exportRules

### PR DESCRIPTION
**Pull request by Haroenv**
in v4 we will rework this under algoliasearch-extras with as much control as needed, similar for browseAll

**Summary**

exportSynonyms and exportRules will get all of the synonyms / rules on an index

**Result**

both implemented naively with the only way to control the flow being specifying hitsPerPage

**Test plan**

```js
index.batchRules([someRules]);
index.searchRules(hitsPerPage?: number).then(saveToFile);
// same for synonyms
```

### CODE REVIEW

**Comment by Haroenv** [comment]
ugh, we don't have enough synonyms on the test app? https://travis-ci.org/algolia/algoliasearch-client-javascript/builds/301357682#L566

File: src/Index.js [hunk 1]
```diff
@@ -745,6 +773,10 @@ Index.prototype.searchRules = function(params, callback) {
   });
 };
 
+Index.prototype.exportRules = function(hitsPerPage) {
```
Comment (vvo) [inline src/Index.js]: Need to pass callback down the line too no? Otherwise here it does not supports callback form.

Comment (Haroenv) [reply src/Index.js]: I was thinking to have this as a "new functions don't have callbacks anymore", but probably better to keep it indeed. Problem is that we use promises internally for this, so you'll still need support (maybe okay with the polyfill we have)

File: src/Index.js [hunk 2]
```diff
@@ -636,6 +636,45 @@ Index.prototype.searchSynonyms = function(params, callback) {
...
+  });
+}
+
+/**
+ * Retrieve all the synonyms in an index
+ * @param [number=100] hitsPerPage The amount of synonyms to retrieve per batch
+ * @param [function] callback will be called after all synonyms are retrieved
+ */
+Index.prototype.exportSynonyms = function(hitsPerPage, callback) {
+  return exportData(this.searchSynonyms, hitsPerPage, callback);
```
Comment (Haroenv) [inline src/Index.js]: It's not exporting 0, 1, 2, and everything is in it twice, and I'm completely lost, @vvo do you mind taking a look?

Comment (Haroenv) [reply src/Index.js]: solved

**Pull request merged**